keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,495 @@
+- Create `test-pod` again
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the pod is evicted 
+
+Test2:
+- Start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled.
+- Create a daemonSet with requiredDuringSchedulingRequiredDuringExecution. 
+- Verify that the pod runs only on nodes that meet the nodeAffinity.
+- Remove the node label.
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+- Add the node's label back.
+- Verify that the daemonSet pod runs on the same node again.
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Check the corresponding metrics(details in SLIs part below).
+- Inspect the `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` configuration.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Events
+  - Event Reason: NodeAffinityEviction
+- [x] API .status
+  - Condition reason: DeletionByNodeAffinityEvictionController 
+- [x] Other (treat as last resort)
+  - Details: Pod with `RequiredDuringSchedulingRequiredDuringExecution` must meet affinity rule during both the scheduling phase and the execution phase.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name:
+    - evictions_node_affinity_total(counts of the eviction/evacuation subresources created by the controller)
+    - pod_failed_eviction_node_affinity_total(counts of the pods that failed to be evicted, labeled by pod name)
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+Two new metric `evictions_node_affinity_total` and `pod_failed_eviction_node_affinity_total` will be added.
+
+### Dependencies
+
```

Comment:
Not exactly a dependency, but adding this is likely to mean changes to node autoscalers, so that they can correctly identify what kind of node to consider provisioning.

Reply:
Should we avoid graduating to beta until at least one in-project autoscaler repo has been updated?

Reply:
> so that they can correctly identify what kind of node to consider provisioning.

That makes senese.
I'm curious how descheduler interact with autoscaler before.

Reply:
I don't follow the relationship.

The cluster-autoscaler would only bring up nodes that match the node affinity. And, in particular, cluster-autoscaler uses the same code as kube-scheduler to determine this. So no changes needed there.

Reply:
I'm suggesting we gate the beta graduation on one of the in project autoscalers updating to match. We can do that gating even if we think that this is likely to happen.

Reply:
It will just happen automatically. But I'm ok with adding it.